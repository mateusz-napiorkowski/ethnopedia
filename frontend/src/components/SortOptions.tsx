import React, { useState, useEffect, useRef } from "react";

export interface Option {
    value: string;
    label: string;
}

interface SortOptionsProps {
    options: Option[];
    sortCategory: string;
    sortDirection: string;
    onSelectCategory: (value: string) => void;
    onSelectDirection: (value: string) => void;
    setCurrentPage: (pageNumber: number) => void;
}

const SortOptions: React.FC<SortOptionsProps> = ({
                                                     options,
                                                     sortCategory,
                                                     sortDirection,
                                                     onSelectCategory,
                                                     onSelectDirection,
                                                     setCurrentPage,
                                                 }) => {
    const [open, setOpen] = useState<boolean>(false);
    const dropdownRef = useRef<HTMLDivElement>(null);
    const buttonRef = useRef<HTMLDivElement>(null);
    const [dropdownWidth, setDropdownWidth] = useState<string>("auto");

    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
                setOpen(false);
            }
        };
        document.addEventListener("mousedown", handleClickOutside);
        return () => document.removeEventListener("mousedown", handleClickOutside);
    }, []);

    useEffect(() => {
        if (buttonRef.current) {
            const longestOption = options.reduce(
                (max, opt) => (opt.label.length > max.length ? opt.label : max),
                ""
            );
            const tempSpan = document.createElement("span");
            tempSpan.style.visibility = "hidden";
            tempSpan.style.position = "absolute";
            tempSpan.style.whiteSpace = "nowrap";
            tempSpan.textContent = longestOption;
            document.body.appendChild(tempSpan);
            setDropdownWidth(`${tempSpan.offsetWidth + 40}px`);
            document.body.removeChild(tempSpan);
        }
    }, [options]);

    const handleSelectCategory = (value: string) => {
        onSelectCategory(value);
        setCurrentPage(1);
        setOpen(false);
    };

    const toggleSortDirection = () => {
        onSelectDirection(sortDirection === "asc" ? "desc" : "asc");
        setCurrentPage(1);
    };

    return (
        <div className="flex items-center space-x-1">
            <div className="relative" ref={dropdownRef}>
                <div
                    ref={buttonRef}
                    onClick={() => setOpen((prev) => !prev)}
                    className="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 dark:border-gray-600 text-sm rounded-lg flex items-center justify-between"
                    style={{ minWidth: dropdownWidth }}
                    aria-label="sortby-category-dropdown"
                >
                    <span>{options.find((opt) => opt.value === sortCategory)?.label || "Wybierz kategorię"}</span>
                    <svg
                        className={`h-4 w-4 ml-2 transition-transform duration-200 ${open ? "rotate-180" : ""}`}
                        fill="none"
                        stroke="currentColor"
                        viewBox="0 0 24 24"
                    >
                        <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
                    </svg>
                </div>
                {open && (
                    <div
                        className="absolute top-full mt-1 z-10 bg-white dark:bg-gray-800 border border-gray-300 dark:border-gray-600 rounded shadow-md max-h-52 overflow-y-auto"
                        style={{ minWidth: dropdownWidth }}
                    >
                        {options.map((option) => (
                            <div
                                key={option.value}
                                onClick={() => handleSelectCategory(option.value)}
                                className="cursor-pointer py-2 px-4 hover:bg-gray-200 dark:hover:bg-gray-700"
                                aria-label={option.label}
                            >
                                {option.label}
                            </div>
                        ))}
                    </div>
                )}
            </div>

            <button
                onClick={toggleSortDirection}
                className="p-2 rounded-md border border-gray-300 dark:border-gray-600 bg-white dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors"
                aria-label="Zmień kierunek sortowania"
                title={sortDirection === "asc" ? "Sortuj rosnąco" : "Sortuj malejąco"}
            >
                <svg
                    className={`h-5 w-5 text-gray-700 dark:text-gray-200 transition-transform duration-200 ${
                        sortDirection === "asc" ? "rotate-180" : ""
                    }`}
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                >
                    <path
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        strokeWidth={2}
                        d="M12 4v16m0 0l6-6m-6 6l-6-6"
                    />
                </svg>


            </button>
        </div>
    );
};

export default SortOptions;
