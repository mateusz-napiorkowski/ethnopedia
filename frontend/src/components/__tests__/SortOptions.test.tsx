import '@testing-library/jest-dom';
import { render } from '@testing-library/react'
import SortOptions from "../SortOptions"
import userEvent from "@testing-library/user-event";

const mockOnSelect = jest.fn()

const user = userEvent.setup()

const renderComponent = () => {
    return render(
        <SortOptions 
            options={[
                {value: "name-asc", label: "Nazwa rosnąco"},
                {value: "name-desc", label: "Nazwa malejąco"}
            ]}
            onSelect={mockOnSelect}
        />
    );
};

describe("Sort options tests", () => {
    beforeEach(() => {
        jest.clearAllMocks()
    });

    it("should render component correctly", () => {           
        const {container} = renderComponent()
        
        expect(container).toMatchSnapshot()
    })

    it("should call onSelect and change selected option in select menu after user chooses an option from select menu", async () => {           
        const { getByRole } = renderComponent()
        const selectElement = getByRole("combobox")
        const nameAscOption = getByRole("option", {
            name: /nazwa malejąco/i
        })
        const nameDescOption = getByRole("option", {
            name: /nazwa rosnąco/i
        })

        await user.selectOptions(selectElement, "name-desc")

        expect(mockOnSelect).toHaveBeenCalled()
        expect((nameAscOption as HTMLOptionElement).selected).toBe(true)
        expect((nameDescOption as HTMLOptionElement).selected).toBe(false)  
    })
})