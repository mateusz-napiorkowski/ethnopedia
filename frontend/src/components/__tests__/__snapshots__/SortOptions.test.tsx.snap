// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort options tests should change selected category when option from dropdown menu when clicked, should call selectCurrentPage(1) 1`] = `
<div>
  <div
    class="flex items-center space-x-1"
  >
    <div
      class="relative"
    >
      <div
        aria-label="sortby-category-dropdown"
        class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 dark:border-gray-600 text-sm rounded-lg flex items-center justify-between"
        style="min-width: 40px;"
      >
        <span>
          Data nagrania
        </span>
        <svg
          class="h-4 w-4 ml-2 transition-transform duration-200 "
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <button
      aria-label="Zmień kierunek sortowania"
      class="p-2 rounded-md border border-gray-300 dark:border-gray-600 bg-white dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors"
      title="Sortuj rosnąco"
    >
      <svg
        class="h-5 w-5 text-gray-700 dark:text-gray-200 transition-transform duration-200 rotate-180"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 4v16m0 0l6-6m-6 6l-6-6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Sort options tests should open dropdown menu when it is clicked 1`] = `
<div>
  <div
    class="flex items-center space-x-1"
  >
    <div
      class="relative"
    >
      <div
        aria-label="sortby-category-dropdown"
        class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 dark:border-gray-600 text-sm rounded-lg flex items-center justify-between"
        style="min-width: 40px;"
      >
        <span>
          Tytuł
        </span>
        <svg
          class="h-4 w-4 ml-2 transition-transform duration-200 rotate-180"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <div
        class="absolute top-full mt-1 z-10 bg-white dark:bg-gray-800 border border-gray-300 dark:border-gray-600 rounded shadow-md max-h-52 overflow-y-auto"
        style="min-width: 40px;"
      >
        <div
          aria-label="Tytuł"
          class="cursor-pointer py-2 px-4 hover:bg-gray-200 dark:hover:bg-gray-700"
        >
          Tytuł
        </div>
        <div
          aria-label="Tytuł.Podtytuł"
          class="cursor-pointer py-2 px-4 hover:bg-gray-200 dark:hover:bg-gray-700"
        >
          Tytuł.Podtytuł
        </div>
        <div
          aria-label="Data nagrania"
          class="cursor-pointer py-2 px-4 hover:bg-gray-200 dark:hover:bg-gray-700"
        >
          Data nagrania
        </div>
      </div>
    </div>
    <button
      aria-label="Zmień kierunek sortowania"
      class="p-2 rounded-md border border-gray-300 dark:border-gray-600 bg-white dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors"
      title="Sortuj rosnąco"
    >
      <svg
        class="h-5 w-5 text-gray-700 dark:text-gray-200 transition-transform duration-200 rotate-180"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 4v16m0 0l6-6m-6 6l-6-6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Sort options tests should render initial state 1`] = `
<div>
  <div
    class="flex items-center space-x-2"
  >
    <div
      class="relative"
    >
      <div
        class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 dark:border-gray-600 text-sm rounded-lg flex items-center justify-between overflow-hidden"
        style="width: 60px;"
      >
        <span
          class="truncate"
          title="Tytuł"
        >
          Tytuł
        </span>
        <svg
          class="h-4 w-4 ml-2 flex-shrink-0 transition-transform duration-200 "
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <button
      class="p-2 rounded-md border border-gray-300 dark:border-gray-600 bg-white dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700 flex items-center justify-center"
      title="Sortowanie rosnące (A → Z)"
    >
      <svg
        class="h-5 w-5 text-gray-700 dark:text-gray-200 transition-transform duration-200 rotate-180"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 4v16m0 0l6-6m-6 6l-6-6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Sort options tests should rotate sorting direction button after sorting direction button is clicked and change its title to 'Sortuj malejąco', should call selectCurrentPage(1) 1`] = `
<div>
  <div
    class="flex items-center space-x-1"
  >
    <div
      class="relative"
    >
      <div
        aria-label="sortby-category-dropdown"
        class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 dark:border-gray-600 text-sm rounded-lg flex items-center justify-between"
        style="min-width: 40px;"
      >
        <span>
          Tytuł
        </span>
        <svg
          class="h-4 w-4 ml-2 transition-transform duration-200 "
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <button
      aria-label="Zmień kierunek sortowania"
      class="p-2 rounded-md border border-gray-300 dark:border-gray-600 bg-white dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors"
      title="Sortuj rosnąco"
    >
      <svg
        class="h-5 w-5 text-gray-700 dark:text-gray-200 transition-transform duration-200 rotate-180"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 4v16m0 0l6-6m-6 6l-6-6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Sort options tests should rotate sorting direction button after sorting direction button is clicked and change its title to 'Sortuj rosnąco', should call selectCurrentPage(1) 1`] = `
<div>
  <div
    class="flex items-center space-x-1"
  >
    <div
      class="relative"
    >
      <div
        aria-label="sortby-category-dropdown"
        class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 dark:border-gray-600 text-sm rounded-lg flex items-center justify-between"
        style="min-width: 40px;"
      >
        <span>
          Tytuł
        </span>
        <svg
          class="h-4 w-4 ml-2 transition-transform duration-200 "
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <button
      aria-label="Zmień kierunek sortowania"
      class="p-2 rounded-md border border-gray-300 dark:border-gray-600 bg-white dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors"
      title="Sortuj malejąco"
    >
      <svg
        class="h-5 w-5 text-gray-700 dark:text-gray-200 transition-transform duration-200 "
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 4v16m0 0l6-6m-6 6l-6-6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;
